### PR TITLE
Copy in order from CallTreeWidget (top-down and bottom-up view)

### PR DESCRIPTION
**Pull request by dpallotti**
While at it, add indentation and indicator of expansion state.

Bug: http://b/177954726

---

Example from top-down view of Trata:
```
+ benchmark (all threads), 100.00% (3730), , , , 
- benchmark [7670], 17.00% (634), , , , 
  - _start, 17.00% (634), 0.00% (0), 100.00%, benchmark, 0x3e6000
    - __libc_start_main, 17.00% (634), 0.00% (0), 100.00%, libc-2.24.so, 0x7f7f6ad731f0
      - main, 17.00% (634), 0.00% (0), 100.00%, benchmark, 0x3e6cd0
        - player::run_ggp(), 17.00% (634), 0.00% (0), 100.00%, benchmark, 0x40ef60
          - player::draw(), 16.97% (633), 0.00% (0), 99.84%, benchmark, 0x40c2b0
            - eva::ui::timeline_player::draw(dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>, eva::profiler&), 15.82% (590), 0.00% (0), 93.21%, benchmark, 0x905b80
              - eva::graphics::player::scene_sequence_viewer::draw(dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>, bool), 15.76% (588), 0.00% (0), 99.66%, benchmark, 0x9784a0
                - eva::graphics::rendering::conga_renderer::render(eva::graphics::rendering::scene_rendering_context&, dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>), 15.66% (584), 0.00% (0), 99.32%, benchmark, 0x9234d0
                  - eva::concurrent::main_thread_task_loop::loop(eva::concurrent::task_sink&, eva::concurrent::thread_group&, bool), 14.96% (558), 0.00% (0), 95.55%, benchmark, 0x9283b0
                      __sched_yield, 8.55% (319), 8.55% (319), 57.17%, libc-2.24.so, 0x7f7f6ae24c70
                    + eva::concurrent::thread_aware_task_sink::execute_task(bool), 6.27% (234), 0.19% (7), 41.94%, benchmark, 0x6aaa00
                      eva::concurrent::thread_aware_task_sink::is_shutting_down() const, 0.13% (5), 0.13% (5), 0.90%, benchmark, 0x6aaca0
                  + eva::graphics::rendering::conga_rendering_tasks::set_inputs(eva::graphics::rendering::scene_rendering_context&, eva::graphics::rendering::rendering_options const&, dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>, bool), 0.46% (17), 0.00% (0), 2.91%, benchmark, 0x933a60
                  + eva::concurrent::dispatching_graph::dispatch(eva::concurrent::task_sink&), 0.16% (6), 0.00% (0), 1.03%, benchmark, 0x924060
                  + eva::graphics::rendering::camera::update(eva::graphics::rendering::view_update_info const&), 0.08% (3), 0.00% (0), 0.51%, benchmark, 0x816150
                + eva::graphics::player::scene_sequence_viewer::find_camera_item(), 0.11% (4), 0.00% (0), 0.68%, benchmark, 0x978570
              + eva::graphics::player::timeline_viewer::draw(eva::graphics::player::time, eva::graphics::player::time, dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>), 0.05% (2), 0.03% (1), 0.34%, benchmark, 0x976e70
            + vulkan_specific::present(), 1.10% (41), 0.00% (0), 6.48%, benchmark, 0x410980
            + vulkan_specific::acquire_next_framebuffer(dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>&, dm_gfx::native_data_handle<dm_gfx::tag_framebuffer_handle>&), 0.03% (1), 0.00% (0), 0.16%, benchmark, 0x4108e0
            + dm_gfx::vulkan::vulkan_gpu_work_pool::get_available_command_buffer(unsigned int), 0.03% (1), 0.00% (0), 0.16%, benchmark, 0x4bbdb0
          + (anonymous namespace)::EventQueueProcessAnyTime(unsigned long, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >), 0.03% (1), 0.00% (0), 0.16%, libggp.so, 0x7f7f6c2bc920
+ benchmark [7680], 12.98% (484), , , , 
+ benchmark [7685], 11.77% (439), , , , 
+ benchmark [7683], 11.58% (432), , , , 
```

### CODE REVIEW

File: src/OrbitQt/CallTreeWidget.cpp [hunk 1]
```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
   column_resizing_state_ = ColumnResizingState::kDone;
 }
 
-static std::string BuildStringFromIndices(const QModelIndexList& indices) {
-  std::string buffer;
+static std::string BuildStringFromIndices(QTreeView* tree_view, const QModelIndexList& indices) {
+  // The order of rows in the selection model is not the order in which the rows are displayed.
+  // We can recover that order by sorting using as comparison function for two rows the
+  // lexicographic order of the vector of the row indices from the root to those rows.
```
Comment (antonrohr) [inline src/OrbitQt/CallTreeWidget.cpp:176]: nit: I understand the code, but I don't understand this sentence. Maybe make 2 or 3 sentences out of it, or alternatively some commas

**Review by antonrohr — APPROVED** [review]
Also tested it, looks great!

File: src/OrbitQt/CallTreeWidget.cpp [hunk 2]
```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
...
+  struct ItemWithAncestorRows {
+    enum ExpandedState { kLeaf = 0, kExpanded, kCollapsed };
+
+    std::string data;
+    bool is_first_in_row = false;
+    ExpandedState expanded_state = kLeaf;
+    std::list<int> ancestor_rows;
```
Comment (beckerhe) [inline src/OrbitQt/CallTreeWidget.cpp:183]: As far as I can see, a forward_list should be enough. But that's kind of nitty...

File: src/OrbitQt/CallTreeWidget.cpp [hunk 3]
```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
...
   }
+
+  // Use stable_sort to maintain the order in the same row (i.e., across columns).
+  std::stable_sort(items.begin(), items.end());
+  std::string buffer;
+  for (size_t i = 0; i < items.size(); ++i) {
+    const ItemWithAncestorRows& item = items[i];
+    if (item.is_first_in_row) {
+      if (i > 0) {
+        buffer.append("\n");
```
Comment (beckerhe) [inline src/OrbitQt/CallTreeWidget.cpp:225]: Nit: You could use the more efficient `std::string::push_back(char)`. But I guess that's mainly pleasing my OCD. So feel free to ignore...

File: src/OrbitQt/CallTreeWidget.cpp [hunk 4]
```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
...
+      }
+
+      for (size_t indent = 0; indent < item.ancestor_rows.size() - 1; ++indent) {
+        buffer.append("  ");
+      }
```
Comment (beckerhe) [inline src/OrbitQt/CallTreeWidget.cpp:230]: What about `buffer.append(2 * item.ancestor_rows.size(); ' ');`?

**Review by beckerhe — APPROVED** [review]
Thanks Davide. I added some minor comments.

File: src/OrbitQt/CallTreeWidget.cpp (hunk 2 above)
Comment (dpallotti) [reply src/OrbitQt/CallTreeWidget.cpp:187]: You had convinced me, but I need `size` below, for the indentation.

File: src/OrbitQt/CallTreeWidget.cpp (hunk 3 above)
Comment (dpallotti) [reply src/OrbitQt/CallTreeWidget.cpp]: Done, I know a thing or two about OCD... :D

File: src/OrbitQt/CallTreeWidget.cpp (hunk 4 above)
Comment (dpallotti) [reply src/OrbitQt/CallTreeWidget.cpp]: Much better. By the way, I'm using `item.ancestor_rows.size() - 1`, excluding the actual row of the element from the count.

File: src/OrbitQt/CallTreeWidget.cpp [hunk 5]
```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
...
+    enum ExpandedState { kLeaf = 0, kExpanded, kCollapsed };
+
+    std::string data;
+    bool is_first_in_row = false;
+    ExpandedState expanded_state = kLeaf;
+    std::list<int> ancestor_rows;
+
+    bool operator<(const ItemWithAncestorRows& that) const {
+      return ancestor_rows < that.ancestor_rows;
+    }
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/OrbitQt/CallTreeWidget.cpp]: You're right, I'm just passing the comparator to `stable_sort` now.

File: src/OrbitQt/CallTreeWidget.cpp (hunk 1 above)
Comment (dpallotti) [reply src/OrbitQt/CallTreeWidget.cpp]: Eh, fair enough. I'm wondering if this comment is of any use at all, but in the meantime I rephrased it, I hope it's understandable now.

File: src/OrbitQt/CallTreeWidget.cpp (hunk 2 above)
Comment (beckerhe) [reply src/OrbitQt/CallTreeWidget.cpp:187]: Ahhh. I see. I wasn't aware that forward_list doesn't have size.

File: src/OrbitQt/CallTreeWidget.cpp (hunk 4 above)
Comment (beckerhe) [reply src/OrbitQt/CallTreeWidget.cpp]: Yes, I missed that! Thanks!

File: src/OrbitQt/CallTreeWidget.cpp (hunk 1 above)
Comment (antonrohr) [reply src/OrbitQt/CallTreeWidget.cpp]: Definitely better! Thanks

**Pull request merged**